Stratis.Bitcoin/Base/ChainHeadersBehavior.cs

```diff
@@ -7,326 +7,390 @@
+            {
+                this.AssertNotAttached();
+                this.chain = value;
+            }
+        }
+
+        private bool invalidHeaderReceived;
+        public bool InvalidHeaderReceived
+        {
+            get
+            {
+                return this.invalidHeaderReceived;
+            }
+        }
+
+        /// <summary>
+        /// Initializes an instanse of the object.
+        /// </summary>
+        /// <param name="chain">Thread safe chain of block headers from genesis.</param>
+        /// <param name="chainState">Information about node's chain.</param>
+        /// <param name="loggerFactory">Factory for creating loggers.</param>
+        public ChainHeadersBehavior(ConcurrentChain chain, ChainState chainState, ILoggerFactory loggerFactory)
+        {
+            Guard.NotNull(chain, nameof(chain));
+
+            this.chainState = chainState;
+            this.chain = chain;
+            this.loggerFactory = loggerFactory;
+            this.logger = loggerFactory.CreateLogger(this.GetType().FullName, $"[{this.GetHashCode():x}] ");
 
-	/// <summary>
-	/// The Chain Behavior is responsible for keeping a ConcurrentChain up to date with the peer, it also responds to getheaders messages.
-	/// </summary>
-	public partial class ChainHeadersBehavior : NodeBehavior
-	{
-		ChainState _State;
-		public ChainHeadersBehavior(ConcurrentChain chain, ChainState chainState)
-		{
-			Guard.NotNull(chain, nameof(chain));
-
-            this._State = chainState;
-			this._Chain = chain;
             this.AutoSync = true;
             this.CanSync = true;
             this.CanRespondToGetHeaders = true;
-		}
-
-		public ChainState SharedState
-		{
-			get
-			{
-				return this._State;
-			}
-		}
-		/// <summary>
-		/// Keep the chain in Sync (Default : true)
-		/// </summary>
-		public bool CanSync
-		{
-			get;
-			set;
-		}
-		/// <summary>
-		/// Respond to getheaders messages (Default : true)
-		/// </summary>
-		public bool CanRespondToGetHeaders
-		{
-			get;
-			set;
-		}
-
-		ConcurrentChain _Chain;
-		public ConcurrentChain Chain
-		{
-			get
-			{
-				return this._Chain;
-			}
-			set
-			{
-				this.AssertNotAttached();
-                this._Chain = value;
-			}
-		}
-
-		int _SynchingCount;
-		/// <summary>
-		/// Using for test, this might not be reliable
-		/// </summary>
-		internal bool Synching
-		{
-			get
-			{
-				return this._SynchingCount != 0;
-			}
-		}
-
-		Timer _Refresh;
-		protected override void AttachCore()
-		{
-            this._Refresh = new Timer(o =>
-			{
-				if (this.AutoSync)
-					this.TrySync();
-			}, null, 0, (int)TimeSpan.FromMinutes(10).TotalMilliseconds);
-			this.RegisterDisposable(this._Refresh);
-			if (this.AttachedNode.State == NodeState.Connected)
-			{
-				var highPoW = this.SharedState.HighestValidatedPoW;
+        }
+
+        protected override void AttachCore()
+        {
+            this.logger.LogTrace("()");
+
+            this.refreshTimer = new Timer(o =>
```

Comment:
I would be happy t remove this timer all together (not not this PR)

Reply:
i think you've mentioned that already somewhere (maybe in 413 discussion?) but i was not sure if this can be removed safely; wouldn't removing this stop us from doing some kind of recovery if something somewhere goes wrong? i don't know, you probably understands this code better, so i would leave that for you, if you want to have this removed, just create an issue for it ...